File: test/skills/test_hf_perceiver.py
# Standard library imports first
import os
import sys
import json
import time
import datetime
import traceback
from unittest.mock import MagicMock, patch

# Third-party imports next
import numpy as np

# Use absolute path setup

# Import hardware detection capabilities if available
try:
    from hardware_detection import (
        HAS_CUDA, HAS_ROCM, HAS_OPENVINO, HAS_MPS, HAS_WEBNN, HAS_WEBGPU,
        detect_all_hardware
    )
    HAS_HARDWARE_DETECTION = True
except ImportError:
    HAS_HARDWARE_DETECTION = False
    # We'll detect hardware manually as fallback
sys.path.insert(0, "/home/barberb/ipfs_accelerate_py")

# Try/except pattern for importing optional dependencies
try:
    import torch
except ImportError:
    torch = MagicMock()
    print("Warning: torch not available, using mock implementation")

try:
    import transformers
except ImportError:
    transformers = MagicMock()
    print("Warning: transformers not available, using mock implementation")

# Import image processing libraries with proper error handling
try:
    from PIL import Image
except ImportError:
    Image = MagicMock()
    print("Warning: PIL not available, using mock implementation")

# Try to import the Perceiver module from ipfs_accelerate_py
try:
    from ipfs_accelerate_py.worker.skillset.hf_perceiver import hf_perceiver
except ImportError:
    # Create a mock class if the real one doesn't exist
    class hf_perceiver:
        def __init__(self, resources=None, metadata=None):
            self.resources = resources if resources else {}
            self.metadata = metadata if metadata else {}
            
        def init_cpu(self, model_name, processor_name, device_label):
            """Mock CPU initialization for Perceiver models"""
            mock_handler = lambda inputs, **kwargs: {
                "logits": np.random.randn(1, 10),
                "predicted_class": "mock_class",
                "implementation_type": "(MOCK)"
            }
            return MagicMock(), MagicMock(), mock_handler, None, 1
            
        def init_cuda(self, model_name, processor_name, device_label):
            """Mock CUDA initialization for Perceiver models"""
            return self.init_cpu(model_name, processor_name, device_label)
    
    print("Warning: hf_perceiver not found, using mock implementation")

class test_hf_perceiver:
    """
    Test class for Hugging Face Perceiver IO models.
    
    The Perceiver IO architecture is a general-purpose encoder-decoder that can handle
    multiple modalities including text, images, audio, video, and multimodal data.
    """
    
    def __init__(self, resources=None, metadata=None):
        """
        Initialize the Perceiver test class.
        
        Args:
            resources (dict, optional): Resources dictionary
            metadata (dict, optional): Metadata dictionary
        """
        # Try to import transformers directly if available
        try:
            import transformers
            transformers_module = transformers
        except ImportError:
            transformers_module = MagicMock()
            
        self.resources = resources if resources else {
            "torch": torch,
            "numpy": np,
            "transformers": transformers_module,
            "PIL": Image
        }
        self.metadata = metadata if metadata else {}
        
        # Create Perceiver instance
        self.perceiver = hf_perceiver(resources=self.resources, metadata=self.metadata)
        
        # Define model variants for different tasks
        self.models = {
            "image_classification": "deepmind/vision-perceiver-conv",
            "text_classification": "deepmind/language-perceiver",
            "multimodal": "deepmind/multimodal-perceiver",
            "masked_language_modeling": "deepmind/language-perceiver-mlm"
        }
        
        # Default to image classification model
        self.default_task = "image_classification"
        self.model_name = self.models[self.default_task]
        self.processor_name = self.model_name  # Usually the same as model name
        
        # Try to validate models
        self._validate_models()
        
        # Create test inputs for different modalities
        self.test_inputs = self._create_test_inputs()
        
        # Initialize collection arrays for examples and status
        self.examples = []
        self.status_messages = {}
        return None
        
    def _validate_models(self):
        """Validate that models exist and fall back if needed"""
        try:
            # Check if we're using real transformers
            if not isinstance(self.resources["transformers"], MagicMock):
                from transformers import AutoConfig
                
                # Try to validate each model
                validated_models = {}
                for task, model in self.models.items():
                    try:
                        print(f"Validating {task} model: {model}")
                        AutoConfig.from_pretrained(model)
                        validated_models[task] = model
                        print(f"✓ Successfully validated {task} model")
                    except Exception as e:
                        print(f"✗ Failed to validate {task} model: {e}")
                        
                # Update models dict with only validated models
                if validated_models:
                    self.models = validated_models
                    self.default_task = list(validated_models.keys())[0]
                    self.model_name = validated_models[self.default_task]
                    self.processor_name = self.model_name
                    print(f"Selected default model: {self.model_name} for {self.default_task}")
                else:
                    print("No models could be validated, using original models")
        except Exception as e:
            print(f"Error validating models: {e}")
            # Keep original models in case of error
    
    def _create_test_inputs(self):
        """Create test inputs for different modalities"""
        test_inputs = {}
        
        # Text input
        test_inputs["text"] = "This is a sample text for testing the Perceiver model."
        
        # Image input
        try:
            # Try to create a test image if PIL is available
            if not isinstance(Image, MagicMock):
                # Create a simple test image (100x100 with gradients)
                width, height = 100, 100
                img = np.zeros((height, width, 3), dtype=np.uint8)
                
                # Create a gradient
                for y in range(height):
                    for x in range(width):
                        img[y, x, 0] = int(255 * y / height)  # R
                        img[y, x, 1] = int(255 * x / width)   # G
                        img[y, x, 2] = int(255 * (x + y) / (width + height))  # B
                
                image = Image.fromarray(img)
                test_inputs["image"] = image
            else:
                test_inputs["image"] = MagicMock()
        except Exception as e:
            print(f"Error creating test image: {e}")
            test_inputs["image"] = MagicMock()
        
        # Multimodal input (text + image)
        test_inputs["multimodal"] = {
            "text": test_inputs["text"],
            "image": test_inputs["image"]
        }
        
        # Audio input (mock)
        test_inputs["audio"] = np.zeros((16000,), dtype=np.float32)  # 1 second at 16kHz
        
        return test_inputs
    
    def _get_test_input_for_task(self, task):
        """Get appropriate test input based on task"""
        if task == "image_classification":
            return self.test_inputs["image"]
        elif task in ["text_classification", "masked_language_modeling"]:
            return self.test_inputs["text"]
        elif task == "multimodal":
            return self.test_inputs["multimodal"]
        else:
            # Default to text
            return self.test_inputs["text"]

    def init_cpu(self, task=None):
        """Initialize Perceiver model on CPU for a specific task"""
        if task is None:
            task = self.default_task
            
        if task not in self.models:
            print(f"Unknown task: {task}, falling back to {self.default_task}")
            task = self.default_task
            
        model_name = self.models[task]
        processor_name = model_name  # Usually the same
        
        print(f"Initializing Perceiver for {task} on CPU with model {model_name}")
        
        # Initialize with CPU
        endpoint, processor, handler, queue, batch_size = self.perceiver.init_cpu(
            model_name,
            processor_name,
            "cpu"
        )
        
        return endpoint, processor, handler, task
    
    def init_cuda(self, task=None):
        """Initialize Perceiver model on CUDA for a specific task"""
        if task is None:
            task = self.default_task
            
        if task not in self.models:
            print(f"Unknown task: {task}, falling back to {self.default_task}")
            task = self.default_task
            
        model_name = self.models[task]
        processor_name = model_name  # Usually the same
        
        print(f"Initializing Perceiver for {task} on CUDA with model {model_name}")
        
        # Check if CUDA is available
        cuda_available = torch.cuda.is_available() if not isinstance(torch, MagicMock) else False
        if not cuda_available:
            print("CUDA not available, falling back to CPU")
            return self.init_cpu(task)
        
        # Initialize with CUDA
        endpoint, processor, handler, queue, batch_size = self.perceiver.init_cuda(
            model_name,
            processor_name,
            "cuda:0"
        )
        
        return endpoint, processor, handler, task
    
    def test_task(self, platform, task=None):
        """Test a specific task on a specific platform"""
        result = {
            "platform": platform,
            "task": task if task is not None else self.default_task,
            "status": "Not run",
            "error": None
        }
        
        try:
            # Initialize model for task
            if platform == "CPU":
                endpoint, processor, handler, task = self.init_cpu(task)
            elif platform == "CUDA":
                endpoint, processor, handler, task = self.init_cuda(task)
            else:
                result["status"] = "Invalid platform"
                result["error"] = f"Unknown platform: {platform}"
                return result
                
            # Get appropriate test input
            test_input = self._get_test_input_for_task(task)
            
            # Test handler
            start_time = time.time()
            output = handler(test_input)
            elapsed_time = time.time() - start_time
            
            # Check if output is valid
            result["output"] = output
            result["elapsed_time"] = elapsed_time
            
            if output is not None:
                result["status"] = "Success"
                
                # Record example
                implementation_type = output.get("implementation_type", "Unknown")
                
                example = {
                    "input": str(test_input)[:100] + "..." if isinstance(test_input, str) and len(str(test_input)) > 100 else str(test_input),
                    "output": output,
                    "timestamp": datetime.datetime.now().isoformat(),
                    "elapsed_time": elapsed_time,
                    "implementation_type": implementation_type,
                    "platform": platform,
                    "task": task
                }
                
                self.examples.append(example)
            else:
                result["status"] = "Failed - No output"
                
        except Exception as e:
            result["status"] = "Error"
            result["error"] = str(e)
            result["traceback"] = traceback.format_exc()
            
        return result

    def test(self):
        """
        Run tests for the Perceiver model across different tasks and platforms.
        
        Returns:
            dict: Structured test results with status, examples, and metadata
        """
        results = {}
        tasks_results = {}
        
        # Test basic initialization
        try:
            results["init"] = "Success" if self.perceiver is not None else "Failed initialization"
        except Exception as e:
            results["init"] = f"Error: {str(e)}"
        
        # Track tested tasks and platforms
        tested_tasks = set()
        tested_platforms = set()
        
        # Run CPU tests for all tasks
        for task in self.models.keys():
            task_result = self.test_task("CPU", task)
            tasks_results[f"cpu_{task}"] = task_result
            tested_tasks.add(task)
            tested_platforms.add("CPU")
            
            # Update status messages
            if task_result["status"] == "Success":
                self.status_messages[f"cpu_{task}"] = "Success"
            else:
                self.status_messages[f"cpu_{task}"] = task_result.get("error", "Failed")
        
        # Run CUDA tests if available
        cuda_available = torch.cuda.is_available() if not isinstance(torch, MagicMock) else False
        if cuda_available:
            for task in self.models.keys():
                task_result = self.test_task("CUDA", task)
                tasks_results[f"cuda_{task}"] = task_result
                tested_platforms.add("CUDA")
                
                # Update status messages
                if task_result["status"] == "Success":
                    self.status_messages[f"cuda_{task}"] = "Success"
                else:
                    self.status_messages[f"cuda_{task}"] = task_result.get("error", "Failed")
        else:
            results["cuda_tests"] = "CUDA not available"
            self.status_messages["cuda"] = "Not available"
        
        # Summarize task results
        for task in tested_tasks:
            cpu_success = tasks_results.get(f"cpu_{task}", {}).get("status") == "Success"
            cuda_success = tasks_results.get(f"cuda_{task}", {}).get("status") == "Success" if cuda_available else False
            
            results[f"task_{task}"] = {
                "cpu": "Success" if cpu_success else "Failed",
                "cuda": "Success" if cuda_success else "Not available" if not cuda_available else "Failed",
                "platforms_success": [p for p in tested_platforms if tasks_results.get(f"{p.lower()}_{task}", {}).get("status") == "Success"]
            }
        
        # Create structured results with tasks details
        structured_results = {
            "status": results,
            "task_results": tasks_results,
            "examples": self.examples,
            "metadata": {
                "models": self.models,
                "default_task": self.default_task,
                "test_timestamp": datetime.datetime.now().isoformat(),
                "python_version": sys.version,
                "torch_version": torch.__version__ if hasattr(torch, "__version__") else "Unknown",
                "transformers_version": transformers.__version__ if hasattr(transformers, "__version__") else "Unknown",
                "platform_status": self.status_messages,
                "cuda_available": cuda_available,
                "cuda_device_count": torch.cuda.device_count() if cuda_available else 0,
                "transformers_mocked": isinstance(self.resources["transformers"], MagicMock)
            }
        }

        return structured_results

    def __test__(self):
        """
        Run tests and compare/save results.
        Handles result collection, comparison with expected results, and storage.
        
        Returns:
            dict: Test results
        """
        # Run actual tests
        test_results = {}
        try:
            test_results = self.test()
        except Exception as e:
            test_results = {
                "status": {"test_error": str(e)},
                "examples": [],
                "metadata": {
                    "error": str(e),
                    "traceback": traceback.format_exc()
                }
            }
        
        # Create directories if they don't exist
        expected_dir = os.path.join(os.path.dirname(__file__), 'expected_results')
        collected_dir = os.path.join(os.path.dirname(__file__), 'collected_results')
        
        os.makedirs(expected_dir, exist_ok=True)
        os.makedirs(collected_dir, exist_ok=True)
        
        # Save collected results
        collected_file = os.path.join(collected_dir, 'hf_perceiver_test_results.json')
        with open(collected_file, 'w') as f:
            json.dump(test_results, f, indent=2, default=str)  # Use default=str to handle non-serializable objects
            print(f"Saved results to {collected_file}")
            
        # Compare with expected results if they exist
        expected_file = os.path.join(expected_dir, 'hf_perceiver_test_results.json')
        if os.path.exists(expected_file):
            try:
                with open(expected_file, 'r') as f:
                    expected_results = json.load(f)
                    
                # Simple check to verify structure
                if "status" in expected_results and "status" in test_results:
                    print("Test completed with expected results structure.")
                else:
                    print("Warning: Results structure does not match expected format.")
            except Exception as e:
                print(f"Error comparing with expected results: {str(e)}")
                # Create expected results file if there's an error
                with open(expected_file, 'w') as f:
                    json.dump(test_results, f, indent=2, default=str)
                    print(f"Created new expected results file: {expected_file}")
        else:
            # Create expected results file if it doesn't exist
            with open(expected_file, 'w') as f:
                json.dump(test_results, f, indent=2, default=str)
                print(f"Created new expected results file: {expected_file}")

        return test_results

if __name__ == "__main__":
    try:
        print("Starting Perceiver test...")
        this_perceiver = test_hf_perceiver()
        results = this_perceiver.__test__()
        print("Perceiver test completed")
        
        # Print test results in detailed format for better parsing
        status_dict = results.get("status", {})
        task_results = results.get("task_results", {})
        examples = results.get("examples", [])
        metadata = results.get("metadata", {})
        
        # Print summary in a parser-friendly format
        print("\nPERCEIVER TEST RESULTS SUMMARY")
        print(f"Default task: {metadata.get('default_task', 'Unknown')}")
        
        # Print task results summary
        print("\nTask Status:")
        for key, value in status_dict.items():
            if key.startswith("task_"):
                task_name = key[5:]  # Remove "task_" prefix
                print(f"  {task_name}:")
                if isinstance(value, dict):
                    for platform, status in value.items():
                        if platform != "platforms_success":
                            print(f"    {platform}: {status}")
                else:
                    print(f"    {value}")
        
        # Print example outputs by task and platform
        task_platform_examples = {}
        
        # Group examples by task and platform
        for example in examples:
            task = example.get("task", "unknown")
            platform = example.get("platform", "Unknown")
            key = f"{task}_{platform}"
            
            if key not in task_platform_examples:
                task_platform_examples[key] = []
                
            task_platform_examples[key].append(example)
        
        # Print one example per task/platform
        print("\nExample Outputs:")
        for key, example_list in task_platform_examples.items():
            if example_list:
                example = example_list[0]
                task = example.get("task", "unknown")
                platform = example.get("platform", "Unknown")
                
                print(f"  {task} - {platform}:")
                print(f"    Input: {example.get('input', 'None')}")
                
                # Format output nicely based on content
                output = example.get("output", {})
                if isinstance(output, dict):
                    # Show only key fields to keep it readable
                    if "logits" in output:
                        print(f"    Output: {output.get('implementation_type', 'Unknown')} - Contains logits")
                    elif "predicted_class" in output:
                        print(f"    Output: Predicted class: {output.get('predicted_class', 'None')}")
                    elif "text" in output:
                        print(f"    Output: {output.get('text', 'None')[:50]}...")
                    else:
                        # Print a short summary
                        keys = list(output.keys())
                        print(f"    Output: Contains {len(keys)} fields: {', '.join(keys)}")
                else:
                    print(f"    Output: {str(output)[:50]}...")
                
                print(f"    Time: {example.get('elapsed_time', 0):.3f}s")
        
    except KeyboardInterrupt:
        print("Tests stopped by user.")
        sys.exit(1)
    except Exception as e:
        print(f"Unexpected error during testing: {str(e)}")
        traceback.print_exc()
        sys.exit(1)

File: ipfs_accelerate_js/test/unit/test_hf_perceiver.ts
// FIXME: Complex template literal
/**;
 * Converted import { {expect, describe: any, it, beforeEach: any, afterEach} from "jest"; } from "Python: test_hf_perceiver.py;"
 * Conversion date: 2025-03-11 04:08:46;
 * This file was automatically converted from Python to TypeScript.;
 * Conversion fidelity might not be 100%, please manual review recommended.;
 */;
";"
import {HfModel} from "src/model/transformers/index/index/index/index/index";"
import {PerceiverConfig} from "src/model/transformers/index/index/index/index/index";"

// WebGPU related imports;
export interface Props {models: prnumber;
  models: prnumber;}
// Standard library imports first;
import * as module; from "*";"
import * as module; from "*";"
import * as module; from "*";"
import * as module; from "*";"
import * as module; from "*";"
import * as module from "*"; import { * as module as np; } from "unittest.mock import * as module, from "*"; patch;"
// Third-party imports next;";"
// Use absolute path setup;
// Import hardware detection capabilities if ((($1) {) {
try ${$1} catch(error) { any): any {HAS_HARDWARE_DETECTION: any: any: any = false;
// We'll detect hardware manually as fallback;'
  sys.path.insert())0, "/home/barberb/ipfs_accelerate_py")}"
// Try/} catch pattern for ((importing optional dependencies {
try ${$1} catch(error) { any) {) { any {torch: any: any: any = MagicMock());
  console.log($1))"Warning: torch !available, using mock implementation")}"
try ${$1} catch(error: any): any {transformers: any: any: any = MagicMock());
  console.log($1))"Warning: transformers !available, using mock implementation")}"
// Import image processing libraries with proper error handling;
try {} catch(error: any): any {Image: any: any: any = MagicMock());
  console.log($1))"Warning: PIL !available, using mock implementation")}"
// Try to import * as module from "*"; Perceiver module from ipfs_accelerate_py;"
}
try ${$1} catch(error: any): any {
// Create a mock class if ((($1) {
  class $1 extends $2 {
    $1($2) {
      this.resources = resources if ($1) {}
      this.metadata = metadata if metadata else {}
      ) {
    $1($2) {
      /** Mock CPU initialization for ((Perceiver models */;
      mock_handler) { any) { any = lambda inputs, **kwargs) { {}
      "logits": np.random.randn())1, 10: any),;"
      "predicted_class": "mock_class",;"
      "implementation_type": "())MOCK)";"
      }
        return MagicMock()), MagicMock()), mock_handler: any, null, 1;
      
    }
    $1($2) {/** Mock CUDA initialization for ((Perceiver models */;
        return this.init_cpu() {)model_name, processor_name) { any, device_label)}
        console.log($1))"Warning) {hf_perceiver !found, using mock implementation")}"
class $1 extends $2 {/** Test class for(Hugging Face Perceiver IO models.}
  The Perceiver IO architecture is a general-purpose encoder-decoder that can handle;
  }
  multiple modalities including text, images { any, audio, video: any, && multimodal data.;
  } */;
  
}
  $1($2) {/** Initialize the Perceiver test class.}
    Args) {
      resources ())dict, optional: any): Resources dictionary;
      metadata ())dict, optional: any): Metadata dictionary */;
// Try to import * as module from "*"; directly if ((($1) {) {"
    try ${$1} catch(error) { any): any {transformers_module: any: any: any = MagicMock());}
    this.resources = resources if ((($1) { ${$1}
      this.metadata = metadata if metadata else {}
// Create Perceiver instance;
      this.perceiver = hf_perceiver())resources=this.resources, metadata) { any) { any: any: any = this.metadata);
// Define model variants for ((different tasks;
    this.models = {}) {"image_classification") { "deepmind/vision-perceiver-conv",;"
      "text_classification": "deepmind/language-perceiver",;"
      "multimodal": "deepmind/multimodal-perceiver",;"
      "masked_language_modeling": "deepmind/language-perceiver-mlm"}"
// Default to image classification model;
      this.default_task = "image_classification";"
      this.model_name = this.models[this.default_task],;
      this.processor_name = this.model_name  # Usually the same as model name;
// Try to validate models;
      this._validate_models());
// Create test inputs for ((different modalities;
      this.test_inputs = this._create_test_inputs() {);
// Initialize collection arrays for examples && status;
      this.examples = [],;
      this.status_messages = {}
      return null;
    
  $1($2) {
    /** Validate that models exist && fall back if ((($1) {
    try {
// Check if ($1) {
      if ($1) {}
// Try to validate each model;
      validated_models) { any) { any = {}
        for (task, model in this.Object.entries($1))) {
          try ${$1} catch(error) { any)) { any {console.log($1))`$1`)}
// Update models dict with only validated models;
        if ((($1) { ${$1} else { ${$1} catch(error) { any)) { any {console.log($1))`$1`)}
// Keep original models in case of error;
  
    }
  $1($2) { */Create test inputs for ((different modalities/** test_inputs) { any) { any: any = {}
// Text input;
    test_inputs["text"], = "This is a sample text for ((testing the Perceiver model.";"
    ,;
// Image input;
    try {
// Try to create a test image if ((($1) {
      if ($1) { ${$1} else { ${$1} catch(error) { any)) { any {console.log($1))`$1`)}
      test_inputs["image"], = MagicMock());"
}
// Multimodal input ())text + image);
      test_inputs["multimodal"], = {},;"
      "text") {test_inputs["text"],;"
      "image") { test_inputs["image"]}"
// Audio input ())mock);
      test_inputs["audio"] = np.zeros())())16000,), dtype: any: any: any = np.float32)  # 1 second at 16kHz;"
      ,;
        return test_inputs;
  
  $1($2) { */Get appropriate test input based on task/** if ((($1) {return this.test_inputs["image"]}"
    } else if (($1) {,;
            return this.test_inputs["text"],;"
    else if (($1) { ${$1} else {// Default to text;
            return this.test_inputs["text"]}"
  $1($2) { */Initialize Perceiver model on CPU for ((a specific task/** if ($1) {
      task) {any = this.default_task;}
    if (($1) {
      console.log($1))`$1`);
      task) {any = this.default_task;}
      model_name) { any) { any) { any = this.models[task],;
      processor_name) {any = model_name  # Usually the same;}
      console.log($1))`$1`);
    
  }
// Initialize with CPU;
      endpoint, processor) { any, handler, queue: any, batch_size: any: any: any = this.perceiver.init_cpu());
      model_name,;
      processor_name: any,;
      "cpu";"
      );
    
      return endpoint, processor: any, handler, task;
  
  $1($2) { */Initialize Perceiver model on CUDA for ((a specific task/** if ((($1) {
      task) {any = this.default_task;}
    if (($1) {
      console.log($1))`$1`);
      task) {any = this.default_task;}
      model_name) {any = this.models[task],;
      processor_name) { any) { any: any = model_name  # Usually the same;}
      console.log($1))`$1`);
// Check if ((CUDA is available;
    cuda_available) { any) { any: any = torch.cuda.is_available()) if ((($1) {
    if ($1) {console.log($1))"CUDA !available, falling back to CPU");"
      return this.init_cpu())task)}
// Initialize with CUDA;
    }
      endpoint, processor) { any, handler, queue: any, batch_size) { any: any: any = this.perceiver.init_cuda());
      model_name,;
      processor_name: any,;
      "cuda:0";"
      );
    
      return endpoint, processor: any, handler, task;
  
  $1($2) { */Test a specific task on a specific platform/** result: any: any = {}
    "platform": platform,;"
      "task": task if ((($1) { ${$1}"
    try {
// Initialize model for ((task;
      if ($1) {
        endpoint, processor) { any, handler, task) { any) {any = this.init_cpu())task);} else if (((($1) { ${$1} else {result["status"] = "Invalid platform",;"
        result["error"] = `$1`,;"
        return result}
// Get appropriate test input;
      }
        test_input) {any = this._get_test_input_for_task())task);}
// Test handler;
        start_time) { any) { any: any = time.time());
        output) { any: any: any = handler())test_input);
        elapsed_time: any: any: any = time.time()) - start_time;
// Check if ((output is valid;
        result["output"] = output,;"
        result["elapsed_time"] = elapsed_time,;"
      ) {
      if (($1) {
        result["status"] = "Success";"
        ,;
// Record example;
        implementation_type) {any = output.get())"implementation_type", "Unknown");}"
        example) { any: any = {}
        "input": str())test_input)[:100] + "..." if ((($1) { ${$1}"
        
        this.$1.push($2))example);
      } else { ${$1} catch(error) { any)) { any {result["status"] = "Error"}"
      result["error"] = str())e),;"
      result["traceback"] = traceback.format_exc());"
      ,;
        return result;

  $1($2) {*/;
    Run tests for ((the Perceiver model across different tasks && platforms.}
    Returns) {
      dict) { Structured test results with status, examples: any, && metadata;
      /** results: any: any = {}
      tasks_results: any: any: any = {}
// Test basic initialization;
    try {
      results["init"] = "Success" if ((($1) { ${$1} catch(error) { any)) { any {results["init"] = `$1`}"
      ,;
// Track tested tasks && platforms;
    }
      tested_tasks: any: any: any = set());
      tested_platforms: any: any: any = set());
// Run CPU tests for ((all tasks;
    for task in this.Object.keys($1) {)) {
      task_result) { any: any = this.test_task())"CPU", task: any);"
      tasks_results[`$1`] = task_result,;
      tested_tasks.add())task);
      tested_platforms.add())"CPU");"
// Update status messages;
      if ((($1) { ${$1} else {this.status_messages[`$1`] = task_result,.get())"error", "Failed")}"
// Run CUDA tests if ($1) {) {
        cuda_available) { any: any = torch.cuda.is_available()) if ((!isinstance() {)torch, MagicMock) { any) else { false;
    if (($1) {
      for ((task in this.Object.keys($1) {)) {
        task_result) {any = this.test_task())"CUDA", task) { any);"
        tasks_results[`$1`] = task_result,;
        tested_platforms.add())"CUDA")}"
// Update status messages;
        if (($1) { ${$1} else { ${$1} else {results["cuda_tests"] = "CUDA !available"}"
      this.status_messages["cuda"] = "Not available";"
      ,;
// Summarize task results;
    for (((const $1 of $2) {
      cpu_success) { any) { any) { any = tasks_results.get())`$1`, {}).get())"status") == "Success";"
      cuda_success) { any: any: any = tasks_results.get())`$1`, {}).get())"status") == "Success" if ((cuda_available else {false}"
      results[`$1`] = {}) {,;
        "cpu") { "Success" if ((($1) {"
        "cuda") { "Success" if (($1) { ${$1}"
// Create structured results with tasks details;
    structured_results) { any) { any = {}:;
      "status": results,;"
      "task_results": tasks_results,;"
      "examples": this.examples,;"
      "metadata": {}"
      "models": this.models,;"
      "default_task": this.default_task,;"
      "test_timestamp": datetime.datetime.now()).isoformat()),;"
      "python_version": sys.version,;"
        "torch_version": torch.__version__ if ((($1) {"
        "transformers_version") { transformers.__version__ if (($1) {"
          "platform_status") { this.status_messages,;"
          "cuda_available") { cuda_available,;"
        "cuda_device_count": torch.cuda.device_count()) if ((($1) { ${$1}"
          return structured_results;

        }
  $1($2) {*/;
    Run tests && compare/save results.;
    Handles result collection, comparison with expected results, && storage.}
    Returns) {
      dict) { Test results;
      """;"
// Run actual tests;
      test_results: any: any = {}
    try ${$1} catch(error: any): any {
      test_results: any: any = {}
      "status": {}"test_error": str())e)},;"
      "examples": [],;"
      "metadata": {}"
      "error": str())e),;"
      "traceback": traceback.format_exc());"
      }
// Create directories if ((they don't exist;'
      expected_dir) { any) { any: any = os.path.join())os.path.dirname())__file__), 'expected_results');'
      collected_dir: any: any: any = os.path.join())os.path.dirname())__file__), 'collected_results');'
    
      os.makedirs())expected_dir, exist_ok: any: any: any = true);
      os.makedirs())collected_dir, exist_ok: any: any: any = true);
// Save collected results;
    collected_file: any: any = os.path.join())collected_dir, 'hf_perceiver_test_results.json'):;'
    with open())collected_file, 'w') as f:;'
      json.dump())test_results, f: any, indent: any: any = 2, default: any: any = str)  # Use default: any: any: any = str to handle non-serializable objects;
      console.log($1))`$1`);
// Compare with expected results if ((they exist;
    expected_file) { any) { any = os.path.join())expected_dir, 'hf_perceiver_test_results.json'):;'
    if ((($1) {
      try {
        with open())expected_file, 'r') as f) {expected_results) { any: any: any = json.load())f);}'
// Simple check to verify structure;
        if ((($1) { ${$1} else { ${$1} catch(error) { any)) { any {console.log($1))`$1`)}
// Create expected results file if ((($1) { ${$1} else {
// Create expected results file if ($1) {
      with open())expected_file, 'w') as f) {}'
        json.dump())test_results, f) { any, indent: any: any = 2, default: any: any: any = str);
        }
        console.log($1))`$1`);

    }
      return test_results;

if ((($1) {
  try {
    console.log($1))"Starting Perceiver test...");"
    this_perceiver) {any = test_hf_perceiver());
    results) { any: any: any = this_perceiver.__test__());
    console.log($1))"Perceiver test completed")}"
// Print test results in detailed format for ((better parsing;
    status_dict) { any) { any: any = results.get())"status", {});"
    task_results: any: any: any = results.get())"task_results", {});"
    examples: any: any: any = results.get())"examples", [],);"
    metadata: any: any: any = results.get())"metadata", {});"
    
}
// Print summary in a parser-friendly format;
    console.log($1))"\nPERCEIVER TEST RESULTS SUMMARY");"
    console.log($1))`$1`default_task', 'Unknown')}");'
// Print task results summary;
    console.log($1))"\nTask Status:");"
    for ((key) { any, value in Object.entries($1) {)) {
      if ((($1) {
        task_name) { any) { any = key[5:]  # Remove "task_" prefix,;"
        console.log($1))`$1`);
        if ((($1) {
          for ((platform) { any, status in Object.entries($1) {)) {
            if (($1) { ${$1} else {console.log($1))`$1`)}
// Print example outputs by task && platform;
      }
          task_platform_examples) { any) { any: any = {}
// Group examples by task && platform;
    for ((const $1 of $2) {
      task) {any = example.get())"task", "unknown");"
      platform) { any: any: any = example.get())"platform", "Unknown");"
      key: any: any: any = `$1`;}
      if ((($1) {task_platform_examples[key] = []}
        task_platform_examples[key].append())example);
        ,;
// Print one example per task/platform;
        console.log($1))"\nExample Outputs) {");"
    for ((key) { any, example_list in Object.entries($1) {)) {
      if (($1) { ${$1}");"
// Format output nicely based on content;
        output) { any) { any: any = example.get())"output", {});"
        if ((($1) {
// Show only key fields to keep it readable;
          if ($1) { ${$1} - Contains logits");"
          else if (($1) { ${$1}");"
          elif ($1) { ${$1}...");"
} else { ${$1}");"
        } else { ${$1}s");"
        } catch(error) { any) ${$1} catch(error) { any)) { any {
    console.log($1))`$1`);
    traceback.print_exc());
    sys.exit())1);